engines/ic3sa.cpp

```diff
@@ -0,0 +1,876 @@
+      conc_ts_(ts, to_prover_solver_),
+      f_unroller_(conc_ts_, 0),  // zero means pure-functional unrolling
+      boolsort_(solver_->make_sort(BOOL))
+{
+  engine_ = Engine::IC3SA_ENGINE;
+}
+
+IC3Formula IC3SA::get_model_ic3formula() const
+{
+  UnorderedTermSet cube_lits;
+  // first populate with predicates
+  for (const auto & p : predset_) {
+    if (solver_->get_value(p) == solver_true_) {
+      cube_lits.insert(p);
+    } else {
+      cube_lits.insert(solver_->make_term(Not, p));
+    }
+  }
+
+  EquivalenceClasses ec;
+  get_equivalence_classes_from_model(ts_.statevars(), ec);
+  construct_partition(ec, cube_lits);
+  IC3Formula cube =
+      ic3formula_conjunction(TermVec(cube_lits.begin(), cube_lits.end()));
+  assert(ic3formula_check_valid(cube));
+
+  return cube;
+}
+
+bool IC3SA::ic3formula_check_valid(const IC3Formula & u) const
+{
+  Sort boolsort = solver_->make_sort(BOOL);
+  // check that children are literals
+  for (const auto & c : u.children) {
+    if (c->get_sort() != boolsort) {
+      return false;
+    }
+  }
+
+  // HACK trouble to accurately identify predicates in boolector
+  // because of simplification. Something that starts as a predicate
+  // might not end up one, e.g. (= ((_ extract 0 0) x) #b0)
+  // just becomes ((_ extract 0 0) x) which we don't consider a predicate
+  // TODO possible fix is to just treat all 1-bit vals as predicates in btor
+  // TODO check if logging
+  if (solver_->get_solver_enum() == BTOR) {
+    return true;
+  }
+
+  Op op;
+  for (auto c : u.children) {
+    op = c->get_op();
+
+    // include bvnot for boolector (if we ever include boolector in this check)
+    if (op == Not || op == BVNot) {
+      c = smart_not(c);
+    }
+
+    // an equality is a predicate, so we just need to check
+    // is_predicate, not specifically for equalities
+    if (!is_predicate(c, boolsort, true)) {
+      return false;
+    }
+  }
+
+  // for now not checking the actual term (e.g. u.term)
+  // it's somewhat hard if the underlying solver does rewriting
+
+  // got through all checks without failing
+  return true;
+}
+
+void IC3SA::predecessor_generalization(size_t i,
+                                       const Term & c,
+                                       IC3Formula & pred)
+{
+  UnorderedTermSet all_coi_symbols = projection_set_;
+  justify_coi(ts_.next(c), all_coi_symbols);
+  assert(all_coi_symbols.size());
+
+  // get rid of next-state variables
+  UnorderedTermSet coi_symbols;
+  for (const auto & tt : all_coi_symbols) {
+    if (!ts_.is_next_var(tt)) {
+      coi_symbols.insert(tt);
+    }
+  }
+  assert(coi_symbols.size() <= ts_.statevars().size());
+  assert(coi_symbols.size());
+
+  logger.log(
+      2,
+      "IC3SA::generalize_predecessor projecting on {}/{} state variables",
+      coi_symbols.size(),
+      ts_.statevars().size());
+
+
+  UnorderedTermSet cube_lits;
+  // first populate with predicates
+  for (const auto & p : predset_) {
```

Comment:
are predicates also going to be added to the cube? 
If so, why?

Reply:
Yeah the predicates are also added to the cube -- it's part of the abstraction as described in the paper. From a theoretical perspective I'm not sure if it's strictly necessary. It might or might not be. But it at least makes it more precise. Unnecessary ones should get removed by predecessor and inductive generalization anyway.

Reply:
predicates can be other than just equality. Right? 
Then, the cubes could contain inequalities (coming from constraints), which is not just equality abstraction.

Reply:
Yeah that’s right. I think that’s why it’s called syntax guided instead of equality abstraction. It’s the predicate assignments plus equalities of other terms.